prov/verbs/src/fi_verbs.h

```diff
@@ -394,33 +443,177 @@ void fi_ibv_monitor_unsubscribe(struct ofi_mem_monitor *notifier,
 				struct ofi_subscription *subscription);
 struct ofi_subscription *fi_ibv_monitor_get_event(struct ofi_mem_monitor *notifier);
 
+/*
+ * An XRC SRQ cannot be created until the associated RX CQ is known,
+ * maintain a list of validated pre-posted receives to post once
+ * the SRQ is created.
+ */
+struct fi_ibv_xrc_srx_prepost {
+	struct slist_entry	prepost_entry;
+	void			*buf;
+	void			*desc;
+	void			*context;
+	size_t			len;
+	fi_addr_t		src_addr;
+};
+
 struct fi_ibv_srq_ep {
 	struct fid_ep		ep_fid;
 	struct ibv_srq		*srq;
+	struct fi_ibv_domain	*domain;
+
+	/* Due to XRC SRQ semantics, an XRC SRX context may only be bound
+	 * to multiple endpoints when they share the same RX CQ. */
+	struct fi_ibv_cq	*srq_cq;
+
+	/* XRC SRQ is not created until endpoint enable */
+	fastlock_t		prepost_lock;
+	ofi_fastlock_acquire_t	pp_fastlock_acquire;
+	ofi_fastlock_release_t	pp_fastlock_release;
```

Comment:
I wouldn't bother trying to optimize the preposting buffer case, assuming this is only handling queuing the posts until the srq has been created.

Reply:
Ok, I have removed the lock optimization.